wikis_parser.py

```diff
@@ -90,95 +93,82 @@ def get_last_update_local_db():
-    If such file doesn't exits, creates it.
-
-    :return: Datetime.datetime of last update or None
-    """
-    if os.path.exists(CSV_UPDATE_TIME):
-        df = pd.read_csv(CSV_UPDATE_TIME)
-        if 'meta' in df.values:
-            update_time = df.loc[df['dbname'] == 'meta', 'update_time'].iloc[0]
-            return update_time
-        return None
-    else:
-        with open(CSV_UPDATE_TIME, "w") as file:
-            file.write('dbname,update_time\n')
-        return None
-
-
-def _update_local_db(update_time):
+def update_local_db(update_time, sources_port=None, user=None, password=None):
     """
     Saves new update time for meta table, creating corresponding row if needed.
 
     :param update_time: Datetime.datetime, time of last update for meta table
+    :param sources_port: Port for connecting to local Sources table through ssh tunneling, if used.
+    :param user: Toolforge username of the tool.
+    :param password: Toolforge password of the tool.
     :return: None
     """
     query = ("insert into Sources(dbname, update_time) values('meta', %s)\n"
              "on duplicate key update update_time = %s"
              )
     try:
-        conn = toolforge.toolsdb(DATABASE_NAME)
+        conn = db_acc.connect_to_user_database(DATABASE_NAME, sources_port, user, password)
         with conn.cursor() as cur:
             time = update_time.strftime('%Y-%m-%d %H:%M:%S')
             cur.execute(query, [time, time])
         conn.commit()
         conn.close()
     except pymysql.err.OperationalError:
-        print('Wikiprojects update checker: failure, please use only in Toolforge environment')
+        print('Wikiprojects update checker: failure, please establish connection to Toolforge')
         exit(1)
 
 
-def update_local_db(update_time):
-    """
-    Saves new update time for meta table, creating corresponding row if needed.
-
-    :param update_time: Datetime.datetime, time of last update for meta table
-    :return: None
-    """
-    df = pd.read_csv(CSV_UPDATE_TIME)
-    if 'meta' in df.values:
-        df.loc[df['dbname'] == 'meta', 'update_time'] = update_time
-        df.to_csv(CSV_UPDATE_TIME, mode='w', header=True, index=False)
-    else:
-        update_time_df = pd.DataFrame([['meta', update_time]], columns=['dbname', 'update_time'])
-        update_time_df.to_csv(CSV_UPDATE_TIME, mode='a', header=False, index=False)
-
-
-def update_checker():
-    wiki_db_update_time = get_creation_date_from_db()
+def update_checker(meta_port=None, sources_port=None, user=None, password=None):
+    wiki_db_update_time = get_creation_date_from_db(meta_port, user, password)
     print('Wikiprojects update checker: time of last update fetched from database')
-    local_db_update_time = get_last_update_local()
-    get_last_update_local_db()
+    local_db_update_time = get_last_update_local_db(sources_port, user, password)
```

Comment:
> I believe the get_db_pages.py file is still to update with argparse?
> 
> For scripts that access multiple dbs, connecting to only metawiki should not work. Should we modify the code such that the entire code can run locally? (even though it's more appropriate to run in grid). Possibly someone can run from other cloud sources, who knows.
> 
> It will be tough to create hundreds of db connections for each db and pass it as arguments. It may be better to connect through shards. See [toolforge:database](https://wikitech.wikimedia.org/wiki/Help:Toolforge/Database#Naming_conventions) and [this](https://wikitech.wikimedia.org/wiki/News/Wiki_Replicas_2020_Redesign#I_have_a_tool_that_queries_all_of_the_wikis_and_I_use_only_one_connection) for more on this.
> In short: we can create 8 connections to access all dbs, but we have to find a way to connect to the appropriate shard for each db.
> 
> Plus I think it would be nice to have a section in readme with commands of connecting to toolforge db. Especially how to connect to replicas, as multiple connection situation might be confusing.
> 
> ps: suggested minor changes to look better from terminal. (with -h flag)
> 
> Update: to test if connecting to metawiki works for all dbs, I ran the db_script locally and it seems to be working so far (still in progress). But I wonder how? I thought we had to connect to each cluster corresponding to that db?



> I believe the get_db_pages.py file is still to update with argparse?
> 
> For scripts that access multiple dbs, connecting to only metawiki should not work. Should we modify the code such that the entire code can run locally? (even though it's more appropriate to run in grid). Possibly someone can run from other cloud sources, who knows.
> 
> It will be tough to create hundreds of db connections for each db and pass it as arguments. It may be better to connect through shards. See [toolforge:database](https://wikitech.wikimedia.org/wiki/Help:Toolforge/Database#Naming_conventions) and [this](https://wikitech.wikimedia.org/wiki/News/Wiki_Replicas_2020_Redesign#I_have_a_tool_that_queries_all_of_the_wikis_and_I_use_only_one_connection) for more on this.
> In short: we can create 8 connections to access all dbs, but we have to find a way to connect to the appropriate shard for each db.
> 
> Plus I think it would be nice to have a section in readme with commands of connecting to toolforge db. Especially how to connect to replicas, as multiple connection situation might be confusing.
> 
> ps: suggested minor changes to look better from terminal. (with -h flag)
> 
> Update: to test if connecting to metawiki works for all dbs, I ran the db_script locally and it seems to be working so far (still in progress). But I wonder how? I thought we had to connect to each cluster corresponding to that db?

Connecting to multiple dbs was something I found working only through tests. I just noticed that when in Toolforge you connect to some database with `sql dbname`, you can choose different database by doing `use new_dbname;`. The same thing works here, I'm not really sure, how exactly, but it looks like mysql makes shard changing automatically somewhere in it's code.

Yes, it's probably makes sense to describe all the steps needed to establish the connection in README, I'm just thinking of it as of another task (it's strange to describe how to open connection to the Toolforge without mentioning that you need an account and how to create it etc., so it's kind of long)

Reply:
Thats interesting! My script is *still* running, I just want to make sure it can connect to all dbs. But yeah, its working. Thanks!

Reply:
> The name `sources_port` seems too specific to the sources table but it's port to the user_db. As in `fetch_content.py` and other files, this can be better named to indicate the database.
> Similarly are certain instances only in this file.

Fixed in 1536898